### PR TITLE
Implementation of AD5683 Driver.

### PR DESCRIPTION
**Pull request by amiclaus**
No-OS Driver compatible with AD5683R/AD5682R/AD5681R/AD5693R/AD5692R/AD5691R.

Signed-off-by: Antoniu Miclaus <antoniu.miclaus@analog.com>

### CODE REVIEW

File: drivers/ad5683/ad5683.c [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+ *                         0 - SPI peripheral was initialized and the
+ *                             device is present.
+*/
+int32_t ad5683_init(struct ad5683_dev **device,
+		    struct ad5683_init_param init_param)
+{
+	struct ad5683_dev *dev;
+	int32_t ret;
+
+	dev = (struct ad5683_dev *)malloc(sizeof(*dev));
```
Comment (commodo) [inline drivers/ad5683/ad5683.c]: would `calloc()`  be a better idea here ?
calloc() guarantees that the memory s initialized to 0; malloc() doesn't
i agree that there are many places where malloc() is used (though calloc()) would be better suited ;

File: drivers/ad5683/ad5683.h [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+
+#define MAX_RESOLUTION             16 // Maximum resolution of the supported devices
+
+/* Power-down operation modes masks */
+#define AD5683_PWRM_NORMAL         0	// Power Mode Normal
+#define AD5683_PWRM_1K             1	// Power-down mode 1kOhm to GND
+#define AD5683_PWRM_100K           2	// Power-down mode 100kOhm to GND
+#define AD5683_PWRM_THREESTATE     3	// Three-State
+
+#define AD5683_DC_EN               1
```
Comment (commodo) [inline drivers/ad5683/ad5683.h]: as a practice : for `<x>_EN` and `<x>_DIS` value pairs, you could just define the behavior of `0 = disable ; non-zero = enable`

**Comment by stefpopa** [comment]
To me, it looks that the support for these devices can be added to the already existing ad5686 driver, as they are very similar, the only obvious difference being the number of bits. This is the way it was done for the Linux driver.

File: drivers/ad5683/ad5683.c [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+	return ad5683_set_shift_reg(dev, AD5683_CMD_WR_CTRL_REG,
+				    AD5683_CTRL_INT_REF(value));
+}
+
+/**
+ * @brief Set DCEN (daisy-chain enable)
+ *
+ * @param dev   - The device structure.
+ * @param value - Enable or disable daisy-chain mode
+ *                Example : 'AD5683_DC_EN' - daisy-chain enable
```
Comment (commodo) [inline drivers/ad5683/ad5683.c]: related to commit https://github.com/analogdevicesinc/no-OS/pull/104/commits/d1914dad852f657681586dd757949bae0c63a60e#r221907052  you could just do  `uint8_t enable`, where 0 disables and non-zero enables

File: drivers/ad5683/ad5683.c [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+ *
+ * @return read_back_data - value read from register.
+*/
+uint32_t ad5683_read_back_register(struct ad5683_dev *dev)
+{
+
+	uint32_t read_back_data = 0;
+	uint32_t offset = MAX_RESOLUTION - chip_info[dev->act_device].resolution;
+
+	if(chip_info[dev->act_device].communication == SPI) {
```
Comment (commodo) [inline drivers/ad5683/ad5683.c]: maybe check if `dev` is NULL before doing anything here with it

File: drivers/ad5683/ad5683.c (hunk 1 above)
Comment (amiclaus) [reply drivers/ad5683/ad5683.c]: The structure is initialized afterwards in the same function with init_params. Does it make sense to initialize it twice? Moreover, I think `malloc` is a bit faster that `calloc`.

File: drivers/ad5683/ad5683.c [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+ * 		       parameters.
+ *
+ * @return ret - The result of the initialization procedure.
+ *               Example: -1 - SPI peripheral was not initialized or the
+ *                             device is not present.
+ *                         0 - SPI peripheral was initialized and the
+ *                             device is present.
+*/
+int32_t ad5683_init(struct ad5683_dev **device,
+		    struct ad5683_init_param init_param)
```
Comment (commodo) [inline drivers/ad5683/ad5683.c]: pass this as pointer ?

File: drivers/ad5683/ad5683.c (hunk 1 above)
Comment (commodo) [reply drivers/ad5683/ad5683.c]: agreed; malloc() is faster ; but calloc() is less error prone ;
i admit i haven't looked at the init of the device, but, if it's fully initialized, and no field member is omitted, then malloc() is fine

**Comment by commodo** [comment]
i would 2nd Stefan's comment:  https://github.com/analogdevicesinc/no-OS/pull/104#issuecomment-426232929
i didn't check to see if there is an existing driver already ;

but it would make sense to extend that if possible

**Comment by amiclaus** [comment]
> To me, it looks that the support for these devices can be added to the already existing ad5686 driver, as they are very similar, the only obvious difference being the number of bits. This is the way it was done for the Linux driver.

The input shift register content is pretty much different from the one in the AD5686. The AD5683 has a Write control register that is accessed via a single command and contains multiple functionalities, while in the AD5686 driver each functionality is accessed via command bits.

**Comment by commodo** [comment]
register addresses can be made configurable per device type ;

we could have a `ad5686_init()` and `ad5683_init()` functions in the same driver, if the drivers differ only by a few elements, but most of the functionality is similar

**Comment by amiclaus** [comment]
Creating new PR with support for these devices in the AD5686 driver.

**Pull request closed without merging**